contracts/TandaPay.sol

```diff
@@ -54,75 +51,101 @@ contract Tandapay {
-        // Use userList to initialize userMapping
+
+        groups.push(newGroup);
+
         Group storage currentGroup = groups[groupIndex];
-        for (uint i = 0; i < currentGroup.userList.length; i++) {
-            currentGroup.userMapping[currentGroup.userList[i]] = 1;
+        for (uint i = 0; i < policyholders.length; i++) {
+          /* initial periodCount is 1
+          users map to 1 to indicate they have to pay premiums for period 1 */
+          currentGroup.userMapping[policyholders[i]] = 1;
         }
         require(currentGroup.userMapping[secretary] == 1); // Checks if secreatry was one of the passed in policyholders
-        
+
         groupIndex += 1;
     }
 
     // A Secretary can initiate a period for policyholders to send in their premiums
-    function startPrePeriod(uint groupId) public secretaryOnly(groupId) validId(groupId) {
+    function startPrePeriod(uint groupId) public secretaryOnly(groupId) {
         Group storage currentGroup = groups[groupId];
-        
+
         currentGroup.oldPeriod = currentGroup.newPeriod;
         currentGroup.newPeriod = Period(Phase.PRE, now);
     }
 
     // A policyholder can send a premium payment to a group
-    function sendPremium(uint groupId) public payable validId(groupId) {
-        Group storage currentGroup = groups[groupId];
-        
-        require(currentGroup.userMapping[msg.sender] == 1); // User is part of group and has not paid the premium
-        require(currentGroup.newPeriod.phase == Phase.PRE); // Assume any Period in pre-period phase is the newPeriod
-        require(msg.value == currentGroup.premium);
-        
-        currentGroup.userMapping[msg.sender] = 2;
-        currentGroup.etherBalance += msg.value;
-        currentGroup.paidPremiumCount += 1;
+    function sendPremium(uint groupId) public payable {
+      Group storage currentGroup = groups[groupId]; //If groupId is not valid, errors here
+      require(msg.value == currentGroup.premium);
+      require(currentGroup.newPeriod.phase == Phase.PRE); // Assume any Period in pre-period phase is the newPeriod
+      require(currentGroup.userMapping[msg.sender] == currentGroup.periodCount); // User is part of group and has not paid the premium
+
+      currentGroup.userMapping[msg.sender]++;
+      currentGroup.etherBalance += msg.value;
+      currentGroup.paidPremiumCount += 1;
     }
 
     // A Secretary can start the active period
-    function startActivePeriod(uint groupId) public secretaryOnly(groupId) validId(groupId){
-        Group storage currentGroup = groups[groupId];
-                
-        require(currentGroup.paidPremiumCount == currentGroup.userList.length); //All premiums have been paid
-        require(currentGroup.newPeriod.phase == Phase.PRE); //Group is not active
-        
-        currentGroup.newPeriod.phase = Phase.ACTIVE;
-        currentGroup.currentPeriod = 2;
-        // Reset userMapping
-        for (uint i = 0; i < currentGroup.userList.length; i++) {
```

Comment:
Why did you delete this part that resets the user mapping? It seems necessary to me

Reply:
I added periodCount so every time a user makes a payment, usermapping increments. so the mapping contains all users and their latest paid period as well. 
So we don't need to keep a separate array for users